libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -716,3 +762,80 @@ void test_MQTT_Subscribe_Publish_With_Qos_2( void )
                        MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
     TEST_ASSERT_TRUE( receivedUnsubAck );
 }
+
+void test_MQTT_Connect_LWT( void )
+{
+    int secondTcpSocket;
+    SSL * pSecondSSLContext;
+    bool sessionPresent;
+    MQTTContext_t secondContext;
+
+    /* Establish a second TCP connection with the server endpoint. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS,
+                       tcpConnectToServer( BROKER_ENDPOINT, BROKER_PORT, &secondTcpSocket ) );
+    TEST_ASSERT_NOT_EQUAL( -1, secondTcpSocket );
+
+    /* Establish TLS connection on top of TCP connection. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS, tlsSetup( secondTcpSocket, SERVER_ROOT_CA_CERT_PATH, &pSecondSSLContext ) );
+    TEST_ASSERT_NOT_NULL( pSecondSSLContext );
+
+    /* Establish MQTT session on top of the TCP+TLS connection. */
+    useLWTClientIdentifier = true;
+    establishMqttSession( &secondContext, pSecondSSLContext, true, &sessionPresent );
+
+    /* Subscribe to LWT Topic. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, subscribeToTopic(
+                           &context, TEST_MQTT_LWT_TOPIC, MQTTQoS0 ) );
+
+    /* Abruptly terminate TLS session. */
+    ( void ) closeTlsSession( pSecondSSLContext );
+
+    /* Terminate TCP connection. */
+    ( void ) closeTcpConnection( secondTcpSocket );
+
+    /* Run the process loop to receive the LWT. Allow some more time for the
+     * server to realize the connection is closed. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Test if we have received the LWT. */
+    TEST_ASSERT_EQUAL( MQTTQoS0, incomingInfo.qos );
+    TEST_ASSERT_EQUAL( TEST_MQTT_LWT_TOPIC_LENGTH, incomingInfo.topicNameLength );
+    TEST_ASSERT_EQUAL_MEMORY( TEST_MQTT_LWT_TOPIC,
+                              incomingInfo.pTopicName,
+                              TEST_MQTT_LWT_TOPIC_LENGTH );
+    TEST_ASSERT_EQUAL( strlen( MQTT_EXAMPLE_MESSAGE ), incomingInfo.payloadLength );
+    TEST_ASSERT_EQUAL_MEMORY( MQTT_EXAMPLE_MESSAGE,
+                              incomingInfo.pPayload,
+                              incomingInfo.payloadLength );
+
+    /* Un-subscribe from a topic with Qos 0. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, unsubscribeFromTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS0 ) );
+
+    /* We expect an UNSUBACK from the broker for the unsubscribe operation. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_TRUE( receivedUnsubAck );
```

Comment:
As the `tearDown()` function uses the first connection's TCP socket and SSL context, the second connection would not be cleaned after the test terminates, and could constitute memory leaks.
You may need to make the TCP, TLS and MQTT connection contexts of the second connection global and update the tear down accordingly.

Reply:
The only situation where the second connection is not cleaned up is when line 787 fails. Do we need to worry about a memory leak when the test is already failing?

Reply:
Never mind, I didn't notice that it is the second connection that is abruptly closed instead of the first one!